dspy/predict/code_act.py

```diff
@@ -0,0 +1,121 @@
+import logging
+import inspect
+
+from typing import Callable, Union, Type
+from inspect import Signature
+
+import dspy
+from dspy.primitives.python_interpreter import PythonInterpreter
+from dspy.primitives.tool import Tool
+from dspy.signatures.signature import ensure_signature
+from dspy.predict.react import ReAct
+from dspy.predict.program_of_thought import ProgramOfThought
+
+logger = logging.getLogger(__name__)
+
+class CodeAct(ReAct, ProgramOfThought):
+    """
+    CodeAct is a module that utilizes the Code Interpreter and predefined tools to solve the problem.
+    """
+
+    def __init__(self, signature: Union[str, Type[Signature]], tools: list[Callable], max_iters=5):
+        """
+        Initializes the CodeAct class with the specified model, temperature, and max tokens.
+
+        Args:
+            signature (Union[str, Type[Signature]]): The signature of the module.
+            tools (list[Callable]): The tool callables to be used. CodeAct only accepts functions and not callable objects.
+            max_iters (int): The maximum number of iterations to generate the answer.
+        
+        Example:
+
+            ```python
+            from dspy.predict import CodeAct
+            def factorial(n):
+                if n == 1:
+                    return 1
+                return n * factorial(n-1)
+
+            act = CodeAct("n->factorial", tools=[factorial])
+            act(n=5) # 120
+            ```
+        """
+        self.signature = ensure_signature(signature)
+        self.max_iters = max_iters
+
+        tools = [t if isinstance(t, Tool) else Tool(t) for t in tools]
+        if any(
+            not inspect.isfunction(tool.func) for tool in tools
+        ):
+            raise ValueError("CodeAct only accepts functions and not callable objects.")
```

Comment:
curious - why do we have this constraint?

Reply:
Since it is hard to define the callable object in the python interpreter. Unlike functions, we need to define class and what parameters are passed to the object initialization, which is not visible.